drivers/regulator/regulator_cp9314.c

```diff
@@ -0,0 +1,388 @@
+		return ret;
+	}
+
+	k_msleep(200);
+
+	return 0;
+}
+
+static int regulator_cp9314_otp_init(const struct device *dev)
+{
+	const struct regulator_cp9314_config *config = dev->config;
+	uint8_t value;
+	int i, ret;
+
+	/**
+	 * The PTE_2 field in the PTE_REG_2 register contains the value representing the OTP
+	 * burned on the CP9314 device. The PTE_2 values in relation to the OTP table names
+	 * are shown below.
+	 *
+	 * OTP-1 = 0x0, OTP-2 = 0x1, OTP-3 = 0x3, OTP-4 = 0x4
+	 */
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, CP9314_REG_PTE_REG_2, &value);
+	if (ret < 0) {
+		return ret;
+	}
+
+	value = FIELD_GET(CP9314_PTE_2_MASK, value);
+
+	ret = i2c_reg_write_byte_dt(&config->i2c, CP9314_REG_CRUS_CTRL, CP9314_CRUS_KEY_UNLOCK);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (value == CP9314_PTE_2_OTP_1) {
+		for (i = 0; i < ARRAY_SIZE(otp_1_patch); i++) {
+			i2c_reg_update_byte_dt(&config->i2c, otp_1_patch[i].reg_addr,
+					       otp_1_patch[i].mask, otp_1_patch[i].value);
+		}
+	}
+
+	if (value <= CP9314_PTE_2_OTP_2) {
+		i2c_reg_update_byte_dt(&config->i2c, CP9314_REG_CFG_9, CP9314_VOUT_PCHG_TIME_CFG_0,
+				       0);
+	}
+
+	return i2c_reg_write_byte_dt(&config->i2c, CP9314_REG_CRUS_CTRL, CP9314_CRUS_KEY_LOCK);
+}
+
+static int regulator_cp9314_init(const struct device *dev)
+{
+	const struct regulator_cp9314_config *config = dev->config;
+	struct regulator_cp9314_data *data = dev->data;
+	uint8_t value;
+	int ret;
+
+	if (!device_is_ready(config->i2c.bus)) {
+		LOG_ERR("Bus device is not ready");
+		return -ENODEV;
+	}
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, CP9314_REG_DEVICE_ID, &value);
+	if (ret < 0) {
+		LOG_ERR("No device found:%d\n", ret);
+		return ret;
+	}
+
+	if (value != CP9314_DEV_ID) {
+		LOG_ERR("Invalid device ID found:0x%x!\n", value);
+		return -ENOTSUP;
+	}
+
+	ret = cp9314_do_soft_reset(dev);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, CP9314_REG_BC_STS_C, &value);
+	if (ret < 0) {
+		return ret;
+	}
+
+	value &= CP9314_CHIP_REV_MASK;
+
+	switch (value) {
+	case CP9314_CHIP_REV_B0:
+		LOG_INF("Found CP9314 REV:0x%x\n", value);
+		ret = regulator_cp9314_b0_init(dev);
+		if (ret < 0) {
+			return ret;
+		}
+		break;
+	default:
+		LOG_ERR("Invalid CP9314 REV:0x%x\n", value);
+		return -ENOTSUP;
+	}
+
+	data->rev_id = value;
+
+	ret = regulator_cp9314_otp_init(dev);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (config->en_pin.port != NULL) {
+		if (!gpio_is_ready_dt(&config->en_pin)) {
+			return -ENODEV;
+		}
+
+		ret = gpio_pin_configure_dt(&config->en_pin, GPIO_OUTPUT_INACTIVE);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	ret = i2c_reg_update_byte_dt(&config->i2c, CP9314_REG_CTRL4, CP9314_FRC_OP_MODE,
+				     CP9314_FRC_OP_MODE);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, CP9314_REG_STS_PIN_ADC_0, &value);
+	if (ret < 0) {
+		return ret;
+	}
+
+	value = FIELD_PREP(CP9314_STS_ADDR_LVL, value);
+
+	switch (value) {
+	case 0x0 ... 0x5:
```

Comment:
This syntax is a gcc extension, so should be avoided

Reply:
ack